apiservice/apiservice.go

```diff
@@ -0,0 +1,268 @@
+package apiservice
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow/protobuf/go/flow/access"
+
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/grpcutils"
+)
+
+func NewFlowAPIService(accessNodeAddressAndPort flow.IdentityList, timeout time.Duration) (*FlowAPIService, error) {
+	accessClients := make([]access.AccessAPIClient, accessNodeAddressAndPort.Count())
+	for i, identity := range accessNodeAddressAndPort {
+		if identity.NetworkPubKey == nil {
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithInsecure(), //nolint:staticcheck
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return nil, err
+			}
+
+			accessClients[i] = access.NewAccessAPIClient(clientRPCConnection)
+		} else {
+			tlsConfig, err := grpcutils.DefaultClientTLSConfig(identity.NetworkPubKey)
+			if err != nil {
+				return nil, err
+			}
+
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return nil, err
+			}
+
+			accessClients[i] = access.NewAccessAPIClient(clientRPCConnection)
+		}
+	}
+
+	ret := &FlowAPIService{}
+	ret.upstream = accessClients
+	ret.LocalCache = nil
+	ret.roundRobin = 0
+	return ret, nil
+}
+
+type FlowAPIService struct {
+	access.UnimplementedAccessAPIServer
+	roundRobin int
+	upstream   []access.AccessAPIClient
+	LocalCache access.AccessAPIServer
+}
+
+func (h *FlowAPIService) Ping(context context.Context, req *access.PingRequest) (*access.PingResponse, error) {
+	if h.LocalCache != nil {
+		return (h.LocalCache).Ping(context, req)
+	}
+	if h.upstream == nil || len(h.upstream) == 0 {
+		return nil, status.Errorf(codes.Unimplemented, "method Ping not implemented")
+	}
+	return h.upstream[h.roundRobin].Ping(context, req)
```

Comment:
Rather than reimplementing these in every endpoint, maybe add a `client()` method something like
```golang
func (h *FlowAPIService) client() (access.AccessAPIClient, error) {
	if h.upstream == nil || len(h.upstream) == 0 {
		return nil, status.Errorf(codes.Unimplemented, "method Ping not implemented")
	}
	
	h.mu.Lock()
	defer h.mu.Unlock()
	
	index := h.roundRobin
	h.roundRobin = (h.roundRobin + 1)%len(h.upstream)
	
	return h.upstream[index]
}

func (h *FlowAPIService) Ping(context context.Context, req *access.PingRequest) (*access.PingResponse, error) {
	client, err := h.client()
	if err != nil {
		return nil, err
	}

	return client.Ping(context, req)
}
```

Reply:
Addressed by: https://github.com/GetElastech/flow-go/commit/b13981732e61d7922227ca402078a76fba45e4b1